File: sop-java-testfixtures/src/main/java/sop/testsuite/AbortOnUnsupportedOptionExtension.java
// SPDX-FileCopyrightText: 2025 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package sop.testsuite;

import org.junit.jupiter.api.extension.ExtensionContext;
import org.junit.jupiter.api.extension.TestExecutionExceptionHandler;
import sop.exception.SOPGPException;

import java.lang.annotation.Annotation;

import static org.junit.jupiter.api.Assumptions.assumeTrue;

public class AbortOnUnsupportedOptionExtension implements TestExecutionExceptionHandler {

    @Override
    public void handleTestExecutionException(ExtensionContext extensionContext, Throwable throwable) throws Throwable {
        Class<?> testClass = extensionContext.getRequiredTestClass();
        Annotation annotation = testClass.getAnnotation(AbortOnUnsupportedOption.class);
        if (annotation != null && throwable instanceof SOPGPException.UnsupportedOption) {
            assumeTrue(false, "Test aborted due to: " + throwable.getMessage());
        }
        throw throwable;
    }
}

File: sop-java-testfixtures/src/main/kotlin/sop/testsuite/AbortOnUnsupportedOptionExtension.kt
// SPDX-FileCopyrightText: 2024 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package sop.testsuite

import org.junit.jupiter.api.Assumptions
import org.junit.jupiter.api.extension.ExtensionContext
import org.junit.jupiter.api.extension.TestExecutionExceptionHandler
import sop.exception.SOPGPException

class AbortOnUnsupportedOptionExtension : TestExecutionExceptionHandler {
    override fun handleTestExecutionException(context: ExtensionContext, throwable: Throwable) {
        val testClass = context.requiredTestClass
        val annotation = testClass.getAnnotation(AbortOnUnsupportedOption::class.java)
        if (annotation != null && SOPGPException.UnsupportedOption::class.isInstance(throwable)) {
            Assumptions.assumeTrue(false, "Test aborted due to: " + throwable.message)
        }
        throw throwable
    }
}
